### PR TITLE
RC Migrations

### LINKED ISSUE #3231: Handle RC migrations
## Is your feature request related to a problem? Please describe 👀
Currently if you install an RC version and initialise - if you then install a later RC which includes database migrations, you'll have to reinitialise the database.

This can take a considerable time (for example, if using cold chain) and so is a frustration for testers.
The release schedule has us creating the RC builds at the end of the two development sprints, however for the upcoming release, the first feature is already done. We have the ability to create an RC build early which allows the QA team to spread the testing workload better. I don't want the difficulty of migration being a blocker in our ability to test effectively.

## Describe the solution you'd like 🎁
Could the RC version be stored, and migrations grouped in batches according to RC version, within the major.minor version migration?

### PR DESCRIPTION
**Pull request by jmbrunskill**
Fixes #3231 

# 👩🏻‍💻 What does this PR do?

Attempts to provide an approach to allow better handling of migrations in RC releases (for testing)
Basically there's a option to create a new `rc_pre_migrate` function that will be run before re-running the migration on an RC build. If this doesn't return an error, the normal version migration will re-run.

The idea of this step is it could run any fixups on existing data, or remove tables/columns etc to allow the main migration to re-run.

Migrations now get a migration context that can be used to create conditional logic with in the main migration.

I'm guessing we'd probably only want to do this if we have something we specifically need to do in an RC build.
Hopefully db changes should be rare.

These RC migrations could be cleaned up after a release.

I've also tried to add some improvements to handling of DEV build migrations, by adding a environment variable to ignore migration errors.

`IGNORE_MIGRATION_ERRORS=1 cargo run`

Figuring this could at least allow us to start say an old branch on a new 

If a database has at some time been through an RC migration, it gets permanently tainted, this is logged at startup, and could be exposed in the UI if needed for testers to see?

`2024-07-04 15:35:06.205047000 [WARN] <repository::migrations:226>:Database has been migrated from Pre Release version, database state might be inconsistent`

## 💌 Any notes for the reviewer?

I won't merge this as is, because it has some random example migrations.

Is there a better way? e.g. Just do point releases for RC builds?

This approach suggests we might want try to write out migrations are re-runable, I think this would be quite nice but isn't easy with SQLite in particular.

If we go with this approach, it might be nice

# 🧪 Testing

Think about things you'd want to be able to update in a RC build, and how easy you think it would be to implement with this approach?

Any other better ideas?
Had thought about making the migrations within the migration as an array rather than standard statements, then each migration `step` could be recorded in the database,and could potentially be applied out of order? Similar to the old diesel migrations.

I do think strictly linking the migrations to the release version might not be the best approach, although it is easy to think about...

# 📃 Documentation

- [ ] **Part of an epic**: documentation will be completed for the feature as a whole
- [X] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: 
  1.
  2.

### CODE REVIEW

**Comment by CarlosNZ** [comment]
Looks good. I think this idea of handling pre-release migrations as a special case, and then integrating them more completely into releases, is a good idea.

This might be overkill and too time-consuming, but I wonder if it would be worth going a step further and requiring all migrations to have an "undo" migration, but only until the next release.

Something like this:
- when developing, if adding a migration, add some additional code to reset the database to the last release version, which would be accessed via a command line option: `DATABASE_RESET=true cargo run` for example. Then, we can be confident that at any stage, on any branch, we can run that command and know we'll get the database back to the state of the last official release
- then at release time, someone goes through and puts all the migrations since the last release into a single block, ensures that they work correctly on a database from the last release (which should be accessible via the RESET command), then deletes all the "undo" code.

Happy to discuss further IRL if we want to agree on a more complete solution. Or maybe it's not enough of a problem to worry too much about?

**Comment by jmbrunskill** [comment]
Thanks Carl, lots of good thoughts here.

> Looks good. I think this idea of handling pre-release migrations as a special case, and then integrating them more completely into releases, is a good idea.

I'm not sure if it's clear from the code but I was thinking we'd aim to make the RC_pre_migrate run before the normal migration. So there should be nothing to move from there to the main migration. After the release, we can simply remove all the `rc_pre_migrate` code.

> This might be overkill and too time-consuming, but I wonder if it would be worth going a step further and requiring all migrations to have an "undo" migration, but only until the next release.
> 
> Something like this:
> 
> * when developing, if adding a migration, add some additional code to reset the database to the last release version, which would be accessed via a command line option: `DATABASE_RESET=true cargo run` for example. Then, we can be confident that at any stage, on any branch, we can run that command and know we'll get the database back to the state of the last official release
> * then at release time, someone goes through and puts all the migrations since the last release into a single block, ensures that they work correctly on a database from the last release (which should be accessible via the RESET command), then deletes all the "undo" code.
> 
> Happy to discuss further IRL if we want to agree on a more complete solution. Or maybe it's not enough of a problem to worry too much about?

Having a rollback function is a common feature for migration tools.
That being said I haven't had much success with this, as no one really tests the rollback function and we'd spend lots of time writing undo migrations that are hardly ever used.

I had an idea, that you could (in dev mode) have an option to save the initialized database at the previous migration version. Then reset if just copy that old database into place. If we made that process easier that could be helpful?

Another approach that kind of helps, is that instead of having one big migration for a version, we could have lots of little migrations, that all get run if they haven't been run yet in order.
That way if I add a migration in my branch and you add one in yours they both get run eventually.
(this is kind of how the diesel migrations used to work)

Doesn't solve the migration itself changing but greatly reduces the conflicts in dev mode.
Was thinking about trying something like that too. Could try to play around with that idea?

**Review by mark-prins — COMMENTED** [review]
looks good to me - thanks James; having this in place would save a lot of pain!

I wondered if there'd be a need for a fn `rc_post_migrate` and haven't come up with anything. I was wondering if we'd ever migrate data - say, rename a column in the 'pre' and copy and delete in the 'post' .. but I don't recall any need for that. We can implement if necessary.

**Comment by jmbrunskill** [comment]
> looks good to me - thanks James; having this in place would save a lot of pain!
> 
> I wondered if there'd be a need for a fn `rc_post_migrate` and haven't come up with anything. I was wondering if we'd ever migrate data - say, rename a column in the 'pre' and copy and delete in the 'post' .. but I don't recall any need for that. We can implement if necessary.

I had a similar thought. Can always add it later?

**Comment by Chris-Petty** [comment]
>Had thought about making the migrations within the migration as an array rather than standard statements, then each migration step could be recorded in the database,and could potentially be applied out of order? Similar to the old diesel migrations.
>
> ...
>
> Another approach that kind of helps, is that instead of having one big migration for a version, we could have lots of little migrations, that all get run if they haven't been run yet in order.
> That way if I add a migration in my branch and you add one in yours they both get run eventually.
> (this is kind of how the diesel migrations used to work)

So we could kinda just disconnect migrations from the semantic versioning? I don't really see why not tbh! Just record the length of the migration array in the database, continue from length if it's changed in next migration?

**Review by Chris-Petty — COMMENTED** [review]
Looks cool! Could certainly make devs/testers lives easier. FWIW Legacy doesn't really do anything special here, rust/OMS is just far more strict about data integrity. Legacy you just have broken data that may or may not affect what you're testing, and often you should revert to an earlier datafile.

I guess the only (release) process is making sure that all the pre-release stuff is cleaned out, and we do have a final testing pass to make the sure the release's migrations are rock solid?

**Comment by jmbrunskill** [comment]
> So we could kinda just disconnect migrations from the semantic versioning? I don't really see why not tbh! Just record the length of the migration array in the database, continue from length if it's changed in next migration?

I don't think it's as simple as that, but we could record an id for each migration and run and id's that don't exist.
The length of the array might be the same if something was removed, or your migration you already ran could be the last in the array so it would try to re-run rather than run the missing one. All sounds doable though!

File: server/repository/src/migrations/version.rs [hunk 1]
```diff
@@ -81,6 +81,10 @@ impl Version {
     pub(crate) fn is_pre_release(&self) -> bool {
         self.pre_release.is_some()
     }
+
+    pub(crate) fn is_equivalent(&self, other: &Self) -> bool {
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply server/repository/src/migrations/version.rs:85]: Could use this, but felt it was nicer to change the API to illustrate to the user that it's not quite semantics as normally `==` implies...

File: server/repository/migrations/postgres/2024-08-02T17-00_migration_enum_keys/up.sql [hunk 2]
```diff
@@ -0,0 +1,3 @@
+ALTER TYPE key_type ADD VALUE IF NOT EXISTS 'DATABASE_MIGRATED_FROM_PRE_RELEASE';
+
+ALTER TYPE key_type ADD VALUE IF NOT EXISTS 'DATABASE_MIGRATION_ERROR';
```
Comment (jmbrunskill) [inline server/repository/migrations/postgres/2024-08-02T17-00_migration_enum_keys/up.sql:3]: I thought this was the cleanest way to get these key types in before the new migration process runs...

**Comment by andreievg** [comment]
Thanks @jmbrunskill, i wanted to try 'by step' migrations, i thought they actual deal with the issue more directly, and i like the way it looks actually: https://github.com/msupply-foundation/open-msupply/pull/4519. I would suggest we go with that way, even though you did a bit of work already, i think fragment migrations are easier to reason about, although I am still on the fence with it, as both ways create a lot of use case and open up large surface area for logical errors, but the demand is high so have to supply : )

**Comment by andreievg** [comment]
Closing in favour of https://github.com/msupply-foundation/open-msupply/pull/4519, nice work James, I think it was very valuable, even though not merged, ended up considering the fragment migration more seriously based on your suggestions and discussions in this PR

**Pull request closed without merging**